rllib/algorithms/algorithm.py

```diff
@@ -670,6 +672,36 @@ def setup(self, config: AlgorithmConfig) -> None:
                     "offline.offline_evaluator::OfflineEvaluator"
                 )
 
+        self.trainer_runner = None
+        if self.config.get("_enable_trainer_runner", False):
+            # TODO (Kourosh/Avnishn) worker should not give us observation/action
+            # space. It should be part of the global config.
```

Comment:
Ideally, yes, but we don't want to force the users to do this:

```
config = PPOConfig().environment("CartPole-v1", observation_space=Box(-1.0, 1.0, (4,), float), action_space=Discrete(2))
```

when the env (at least once constructed) already reveals that information automatically. Ideally, the env class itself would already carry the spaces infos, but that's not the case. You have to construct an env object first to see what the spaces are.

Reply:
just to clarify, global config does not mean input config. It can be something inferred at the time of algorithm construction. Basically the preprocessor logic that infers the final obs / action space that the model should be built on should be accessible outside the context of rollout_workers, so that they can be properly re-used else where without needing to have a reference to workers.

Reply:
Yeah, that's correct.
Inferring this during run-time sounds great!
Just want to explain this here for other readers: Because by default, our Algorithm driver process does NOT have any env object (on the "local worker"), we need to get this information from one of the remote workers (who are the only items that carry an actual env object). After accessing the spaces information via ray.get from one of the remote workers, yes, we should store it properly inside Algorithm properties.
Does that make sense?